parser/parser/cc/grammars/typedruby.ypp

```diff
@@ -523,27 +523,14 @@ int yylex(parser::semantic_type *lval, ruby_parser::location *lloc, ruby_parser:
 // bunch of the already-parsed program.
 //
 // So in rules where we expect something like that to happen, we can call this
-// function to fix up the lookahead token by manually creating a tNL
-int maybe_inject_tNL_impl(parser::symbol_type &yyla, ruby_parser::TYPEDRUBY27 &driver, size_t prev_end, size_t next_begin) {
-  // This function is super hacky, and is best understood by carefully reading
-  // yylex and where yylex is called inside the generated parser
-  auto fake_token = driver.lex.tok_view_from_offsets(prev_end, next_begin);
-  auto nl_start = fake_token.find('\n');
-  if (nl_start != std::string_view::npos) {
-    token_t lookahead = yyla.value.token;
-    yyla.clear(); // same as yyclearin macro
-    auto nl_type = ruby_parser::token_type::tNL;
-    auto nl_tok = driver.lex.unadvance(lookahead, nl_type, nl_start, nl_start + 1, "\n");
-    yyla.value.token = nl_tok;
-    driver.last_token = nl_tok;
-    return static_cast<int>(nl_tok->type());
-  } else {
-    // no-op
-    return static_cast<int>(yyla.value.token->type());
-  }
-}
-
-#define maybe_inject_tNL(prev_end, next_begin) ((yyla.type) = yytranslate_(maybe_inject_tNL_impl(yyla, driver, (prev_end), (next_begin))))
+// function to request that the lexer start again after `tDOT` token in the
+// expr_end state.
+#define rewind_and_reset(newPos)     \
+  do {                               \
+    yyclearin;                       \
+    driver.lex.rewind(newPos);       \
+    driver.lex.set_state_expr_end(); \
```

Comment:
The lexer maintains a, um, lot of state; is this going to be sufficient to reset all of it?  (Actually, the driver does too, but I guess that is a separate question...maybe?)

Reply:
Yeah, I'm a little concerned about this too. If it happens to work in places where this PR uses it I think it's fine, but since it doesn't touch other parts of the state it feels like it could be a footgun if someone tries to call it without awareness of its limitations. (For one example: it doesn't reset `in_kwarg`, so if we were to rewind past the beginning of the kwargs, maybe that could screw up... something?) Might make sense to insert a caution in the comments.

Reply:
This is a good point, but I think that luckily this will only happen during error recovery. So it'll be bugs in our new error recovery code, instead of regressions in the existing parser things. That's a tradeoff I'm mostly fine with at the moment, because we we get bug reports for the parser it should hopefully be easy to triage and fix resetting other bits of state that we'll need.

That being said, this is a good callout, and i'll go ahead and audit all the state we're setting now and see how much of it we'll need to manually reset.

Reply:
> So it'll be bugs in our new error recovery code, instead of regressions in the existing parser things. That's a tradeoff I'm mostly fine with at the moment, because we we get bug reports for the parser it should hopefully be easy to triage and fix resetting other bits of state that we'll need.

Yup, I think that's a good argument for trying it and seeing how it goes!

Reply:
I audited the the stateful things and concluded that none of them except for `in_kwarg` looked like they obviously needed to or should be reset.

I moved the logic into a proper helper function where the name makes it clear that it's going to reset the state.

I also limited it to just resetting state to the `expr_end` state, which I think at least makes it clear that the method doesn't attempt to arbitrarily rewind—I don't think we'll need it to, and i think that rewinding to `expr_end` in all the current and foreseeable cases is straightfoward.

Thanks again for the suggestion.